src/Lecture4.hs

```diff
@@ -177,196 +73,76 @@ data Stats = Stats
-non-empty list of rows.
-
-We can use the 'NonEmpty' data type from the 'Data.List.NonEmpty'
-module for this purpose. 'NonEmpty' is like 'List1' from Lecture 3
-exercises (remember that type?) but with the different constructor.
-
-Have a look at the 'sconcat' function from the 'Semigroup' typeclass to
-implement the next task.
--}
+rowToStats (r@(Row p tt c)) 
+    | tt == Buy = Stats { statsTotalPositions = Sum 1, statsTotalSum = Sum (-c), statsAbsoluteMax = Max (c), statsAbsoluteMin = Min (c), statsSellMax = Nothing, statsSellMin = Nothing, statsBuyMax = Just (Max (c)), statsBuyMin = Just (Min (c)), statsLongest = MaxLen (p)}
+	| tt == Sell = Stats { statsTotalPositions = Sum 1, statsTotalSum = Sum (c), statsAbsoluteMax = Max (c), statsAbsoluteMin = Min (c), statsSellMax = Just (Max (c)), statsSellMin = Just (Min (c)), statsBuyMax = Nothing, statsBuyMin = Nothing, statsLongest = MaxLen (p)}
 
 combineRows :: NonEmpty Row -> Stats
-combineRows = error "TODO"
-
-{-
-After we've calculated stats for all rows, we can then pretty-print
-our final result.
-
-If there's no some value (for example, there were not "Buy" products),
-you can return string "no value"
--}
+combineRows (r:|[]) = rowToStats r
+combineRows (r:|rs) = let (s:|st) = head (fmap rowToStats (r:rs)):|tail (fmap rowToStats (r:rs)) in sconcat (s:|st)
 
 displayStats :: Stats -> String
-displayStats = error "TODO"
-
-{-
-Now, we definitely have all the pieces in places! We can write a
-function that takes content of the file (the full content with multiple lines)
-and converts it to pretty-printed stats.
-
-The only problem here is that after parsing a file we might end with
-an empty list of rows but our 'combineRows' function requires to have
-a non-empty list. In that case, you can return a string saying that
-the file doesn't have any products.
-
-🕯 HINT: Ideally, the implementation of 'calculateStats' should be just a
-   composition of several functions. Use already implemented functions, some
-   additional standard functions and maybe introduce helper functions if you need.
-
-🕯 HINT: Have a look at 'mapMaybe' function from 'Data.Maybe' (you may need to import it).
--}
-
+displayStats st = "Total positions:       : "++show (getSum (statsTotalPositions st)) ++"\n" ++"Total final balance    : "++ show (getSum(statsTotalSum st))++"\n" ++"Biggest absolute cost  : "++ show (getMax(statsAbsoluteMax st))++"\n" ++"Smallest absolute cost : "++ show (getMin(statsAbsoluteMin st))++"\n" ++"Max earning            : "++ fromMaybe "no value" (fmap show (fmap getMax(statsSellMax st)))++"\n" ++"Min earning            : "++ fromMaybe "no value" (fmap show (fmap getMin(statsSellMin st)))++"\n" ++"Max spending           : "++ fromMaybe "no value" (fmap show (fmap getMax(statsBuyMax st)))++"\n" ++"Min spending           : "++ fromMaybe "no value" (fmap show (fmap getMin(statsBuyMin st)))++"\n" ++"Longest product name   : "++ unMaxLen(statsLongest st)
```

Comment:
It would be better for readability to write this on separate lines 👾 

Additionally, you can use the standard function `unlines` and put all lines as separate elements of the list. With this approach you also don't need to worry about putting `\n` 🙂